File: SortVision/public/code/radix/php/radixSort.php
<?php
/**
 * Radix Sort Implementation
 * @param array &$arr Array to sort (passed by reference)
 * @return array Sorted array
 */
function radixSort(&$arr) {
    // Find the maximum number to know the number of digits
    $max = getMax($arr);
    
    // Do counting sort for every digit. exp is 10^i where i is current digit number
    for ($exp = 1; $max / $exp >= 1; $exp *= 10) {
        countingSort($arr, $exp);
    }
    return $arr;
}

/**
 * Counting sort for radix sort
 * @param array &$arr Array to sort
 * @param int $exp Current exponent (1, 10, 100, ...)
 */
function countingSort(&$arr, $exp) {
    $n = count($arr);
    $output = array_fill(0, $n, 0); // Output array
    $count = array_fill(0, 10, 0);  // Count array for digits 0-9

    // Store count of occurrences in count[]
    for ($i = 0; $i < $n; $i++) {
        $index = (int)(($arr[$i] / $exp) % 10);
        $count[$index]++;
    }

    // Change count[i] so that count[i] contains actual position of this digit in output[]
    for ($i = 1; $i < 10; $i++) {
        $count[$i] += $count[$i - 1];
    }

    // Build the output array (stable sort)
    for ($i = $n - 1; $i >= 0; $i--) {
        $index = (int)(($arr[$i] / $exp) % 10);
        $output[$count[$index] - 1] = $arr[$i];
        $count[$index]--;
    }

    // Copy the output array to arr[], so that arr[] now contains sorted numbers according to current digit
    for ($i = 0; $i < $n; $i++) {
        $arr[$i] = $output[$i];
    }
}

/**
 * Get maximum value in array
 * @param array $arr Input array
 * @return int Maximum value
 */
function getMax($arr) {
    // Initialize max as the first element
    $max = $arr[0];
    // Traverse array elements to find the maximum
    foreach ($arr as $value) {
        if ($value > $max) {
            $max = $value;
        }
    }
    return $max;
}
?>

File: SortVision/public/code/radix/ruby/radixSort.rb
# radixSort.rb
# 📌 Radix Sort Implementation in Ruby
# 🔄 Non-comparative, integer sorting algorithm

# ----------------------------------------
# 🧠 Get Maximum Value from Array
# Returns the maximum number in the array
# Time Complexity: O(n)
def get_max(arr)
  arr.max || 0
end

# ----------------------------------------
# 🧮 Counting Sort (used as subroutine by Radix Sort)
# Sorts the array based on the digit represented by exp
# Time Complexity: O(n)
# Space Complexity: O(n + k), where k = base (10)
def counting_sort(arr, exp)
  n = arr.length
  output = Array.new(n, 0)     # output array
  count = Array.new(10, 0)     # digit counts for base 10

  # Store count of occurrences in count[]
  arr.each do |num|
    index = (num / exp) % 10
    count[index] += 1
  end

  # Change count[i] so that count[i] now contains actual
  # position of this digit in output[]
  (1...10).each do |i|
    count[i] += count[i - 1]
  end

  # Build the output array (stable sort)
  (n - 1).downto(0) do |i|
    index = (arr[i] / exp) % 10
    output[count[index] - 1] = arr[i]
    count[index] -= 1
  end

  # Copy the output array to arr[]
  (0...n).each do |i|
    arr[i] = output[i]
  end

  arr
end

# ----------------------------------------
# 🚀 Radix Sort Function
# Time Complexity: O(d * (n + b)) where:
#   - d = number of digits
#   - b = base (10)
#   - n = number of elements
# Space Complexity: O(n)
def radix_sort(arr)
  return arr if arr.length <= 1

  max_num = get_max(arr)

  exp = 1
  while max_num / exp > 0
    counting_sort(arr, exp)
    exp *= 10
  end

  arr
end

# ----------------------------------------
# 🧪 Example Usage and Test Cases
if __FILE__ == $0
  puts "📘 Radix Sort in Ruby"
  
  test_cases = [
    [],
    [5],
    [170, 45, 75, 90, 802, 24, 2, 66],
    [3, 3, 3],
    [100, 10, 1],
    [999, 888, 777, 111]
  ]

  test_cases.each_with_index do |test, idx|
    puts "\n🔢 Test Case #{idx + 1}:"
    puts "Input:  #{test}"
    sorted = radix_sort(test.dup)
    puts "Sorted: #{sorted}"
  end
end

# ----------------------------------------
# 📈 Notes:
# - Radix Sort only works efficiently for non-negative integers.
# - Not suitable for negative numbers without modifications.
# - It's faster than comparison sorts (like merge or quick sort)
#   for large datasets with small key sizes.
